runtime-modules/storage/src/lib.rs

```diff
@@ -3493,6 +3498,73 @@ impl<T: Trait> DataObjectStorage<T> for Module<T> {
             .map(|x| x.0)
             .collect()
     }
+    #[allow(clippy::redundant_closure)] // doesn't work with Substrate storage functions.
+    fn funds_needed_for_upload(params: UploadParameters<T>) -> Result<BalanceOf<T>, DispatchError> {
+        // size check:
+        ensure!(
+            !params.object_creation_list.is_empty(),
+            Error::<T>::NoObjectsOnUpload
+        );
+
+        // ensure specified data fee == storage data fee
+        ensure!(
+            params.expected_data_size_fee == DataObjectPerMegabyteFee::<T>::get(),
+            Error::<T>::DataSizeFeeChanged,
+        );
+
+        let object_deletion_prize = Self::data_object_deletion_prize_value();
+        // ensure data object deletion prize
+        ensure!(
+            params.expected_data_object_deletion_prize == object_deletion_prize,
+            Error::<T>::DataObjectDeletionPrizeChanged,
+        );
+
+        let bag_op = BagOperation::<T> {
+            bag_id: params.bag_id.clone(),
+            params: BagOperationParams::<T>::Update(
+                params.object_creation_list,
+                Default::default(),
+            ),
+        };
+
+        let Bag::<T> {
+            stored_by: _,
+            distributed_by: _,
+            deletion_prize,
+            objects_total_size,
+            objects_number,
+        } = Self::ensure_bag_exists(&bag_op.bag_id)?;
+
+        // check and generate any objects to add/remove from storage
+        let object_creation_list = Self::construct_objects_to_upload(&bag_op)?;
+
+        // new candidate voucher
+        let new_voucher_update = VoucherUpdate {
+            objects_total_size,
+            objects_number,
+        }
+        .add_objects_list(object_creation_list.as_slice());
+
+        let init_net_prize =
+            deletion_prize.map_or(Default::default(), |dp| NetDeletionPrize::<T>::Pos(dp));
+
+        let net_prize = match &bag_op.params {
+            BagOperationParams::<T>::Update(creation_list, _) => {
+                Self::compute_net_prize(init_net_prize, creation_list.len(), 0)
+            }
+            _ => Default::default(),
+        };
+
+        // storage fee: zero if VoucherUpdate is default
+        let storage_fee = Self::calculate_data_storage_fee(new_voucher_update.objects_total_size);
+        //Only the case where the net deletion prize (NDP) is positive will be returned, as the
+        //function will be used to validate whether the actor has enough balance to update the assets.
+        let funds_needed_for_upload = match net_prize.add_balance(storage_fee) {
```

Comment:
Won't `net_prize` include old deletin prize? If so, wouldn't `funds_needed_for_upload` in fact be the new total cost, rather than cost of doing this additional upload, which is what we care about?

Reply:
When we call `upload_data_objects` this function will call `try_mutating_storage_state` with 
`BagOperationParams::Update(object_creation_list)`
then 

1. we will ensure bag exists and bag has a deletion prize, that  is some for dynamic bags or none for static ones.
2. Construct objects to upload  where deletion prize is the old one
3. we compute net prize init_value + (deletion_prize * num_obj_to_create)
4. We add net_prize to storage_fee

 So net prize include deletion prize in the calculation, but that is what `try_mutating_storage_state` is doing to do the ensure_funds validation 
I think the goal here is to separate funds validation from try_mutate_storage, right?